lib/streamlit/runtime/memory_uploaded_file_manager.py

```diff
@@ -0,0 +1,135 @@
+    """
+
+    def __init__(self, upload_endpoint: str):
+        self.file_storage: Dict[str, Dict[str, UploadedFileRec]] = defaultdict(dict)
+        self.endpoint = upload_endpoint
+
+    def get_files(
+        self, session_id: str, file_ids: Sequence[str]
+    ) -> List[UploadedFileRec]:
+        """Return a  list of UploadedFileRec for a given sequence of file_ids.
+
+        Parameters
+        ----------
+        session_id
+            The ID of the session that owns the files.
+        file_ids
+            The sequence of ids associated with files to retrieve.
+
+        Returns
+        -------
+        List[UploadedFileRec]
+            A list of URL UploadedFileRec instances, each instance contains information
+            about uploaded file.
+        """
+        session_storage = self.file_storage[session_id]
+        file_recs = []
+
+        for file_id in file_ids:
+            file_rec = session_storage.get(file_id, None)
+            if file_rec is not None:
+                file_recs.append(file_rec)
+
+        return file_recs
+
+    def remove_session_files(self, session_id: str) -> None:
+        """Remove all files associated with a given session."""
+        self.file_storage.pop(session_id, None)
+
+    def __repr__(self) -> str:
+        return util.repr_(self)
+
+    def add_file(
+        self,
+        session_id: str,
+        file: UploadedFileRec,
+    ) -> None:
+        """
+        Safe to call from any thread.
+
+        Parameters
+        ----------
+        session_id
+            The ID of the session that owns the file.
+        file
+            The file to add.
+        """
+
+        self.file_storage[session_id][file.file_id] = file
+
+    def remove_file(self, session_id, file_id):
+        """Remove file with given file_id associated with a given session."""
+        session_storage = self.file_storage[session_id]
+        session_storage.pop(file_id, None)
+
+    def get_upload_urls(
+        self, session_id: str, file_names: Sequence[str]
+    ) -> List[UploadFileUrlInfo]:
+        """Return a list of UploadFileUrlInfo for a given sequence of file_names."""
+        result = []
+        for _ in file_names:
+            file_id = str(uuid.uuid4())
+            result.append(
+                UploadFileUrlInfo(
+                    file_id=file_id,
+                    upload_url=f"{self.endpoint}/{session_id}/{file_id}",
+                    delete_url=f"{self.endpoint}/{session_id}/{file_id}",
+                )
+            )
+        return result
+
+    def get_stats(self) -> List[CacheStat]:
+        """Return the manager's CacheStats.
+
+        Safe to call from any thread.
+        """
+        # Flatten all files into a single list
+        all_files: List[UploadedFileRec] = []
+        # Make copy of self.file_storage for thread safety, to be sure
+        # that main storage won't be changed form other thread
+        file_storage_copy = self.file_storage.copy()
```

Comment:
yes, now we create a duplicate to collect statistics. 
The good news is that this duplicate should be short-living, and Python Garbage Collector should collect it very fast because we don't keep any references to `file_storage_copy`, but yes, this could lead to memory usage increase in case of a lot of stored uploaded files.

Reply:
Maybe you can add a TODO comment or something there, so that we double check this once we do more with the stats.